src/kvstore/CompactionFilter.h

```diff
@@ -0,0 +1,98 @@
+/* Copyright (c) 2019 vesoft inc. All rights reserved.
+ *
+ * This source code is licensed under Apache 2.0 License,
+ * attached with Common Clause Condition 1.0, found in the LICENSES directory.
+ */
+
+#ifndef KVSTORE_COMPACTIONFILTER_H_
+#define KVSTORE_COMPACTIONFILTER_H_
+
+#include "base/Base.h"
+#include <rocksdb/compaction_filter.h>
+#include "base/NebulaKeyUtils.h"
+#include "time/TimeUtils.h"
+#include "kvstore/Common.h"
+
+namespace nebula {
+namespace kvstore {
+
+class KVCompactionFilter final : public rocksdb::CompactionFilter {
+public:
+    KVCompactionFilter(GraphSpaceID spaceId, std::unique_ptr<KVFilter> kvFilter)
+        : spaceId_(spaceId)
+        , kvFilter_(std::move(kvFilter)) {
+    }
+
+    bool Filter(int,
+                const rocksdb::Slice& key,
+                const rocksdb::Slice& val,
+                std::string*,
+                bool*) const override {
+        return kvFilter_->filter(spaceId_,
+                                 folly::StringPiece(key.data(), key.size()),
+                                 folly::StringPiece(val.data(), val.size()));
+    }
+
+    const char* Name() const override {
+        return "KVCompactionFilter";
+    }
+
+private:
+    GraphSpaceID spaceId_;
+    std::unique_ptr<KVFilter> kvFilter_;
+};
+
+class KVCompactionFilterFactory : public rocksdb::CompactionFilterFactory {
+public:
+    KVCompactionFilterFactory() = default;
+
+    virtual ~KVCompactionFilterFactory() = default;
+
+    /**
+     * We should call construct explicitly becasue the instacne is passed by modules outside.
+     * */
+    void construct(GraphSpaceID spaceId, int32_t customFilterIntervalSecs) {
+        spaceId_ = spaceId;
+        customFilterIntervalSecs_ = customFilterIntervalSecs;
+    }
+
+    std::unique_ptr<rocksdb::CompactionFilter>
+    CreateCompactionFilter(const rocksdb::CompactionFilter::Context& context) override {
+        auto now = time::TimeUtils::nowInSeconds();
+        if (context.is_full_compaction) {
+            LOG(INFO) << "Do full compaction!";
+            lastRunCustomFilterTimeSec_ = now;
+            return std::unique_ptr<rocksdb::CompactionFilter>(
+                            new KVCompactionFilter(spaceId_, createKVFilter()));
```

Comment:
Do we have to consider default full compaction without filter?

Reply:
One important purpose for full compaction is to remove useless data.  So we should use custom filter.